minimal_memory_lock/minimal_memory_lock.cpp

```diff
@@ -0,0 +1,273 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <sys/types.h>
+#include <sys/syscall.h>
+#include <sys/mman.h>
+
+#include <stdlib.h>
+#include <limits.h>
+#include <malloc.h>
+#include <sys/resource.h>
+#include <string.h>
+#include <unistd.h>
+#include <pthread.h>
+
+#include <rcutils/cmdline_parser.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+static const char * OPTION_LOCK_MEMORY = "--lock-memory";
+static const char * OPTION_MEMORY_PREALLOCATION = "--memory-preallocation";
+static const char * OPTION_THREAD_STACK_SIZE = "--stacksize";
+static const char * OPTION_USE_SLEEP = "--sleep";
+
+namespace
+{
+void print_usage()
+{
+  printf("\nUsage:\n");
+  printf("%s : lock memory\n", OPTION_LOCK_MEMORY);
+  printf("%s : preallocate memory for the process (MB)\n", OPTION_MEMORY_PREALLOCATION);
+  printf("%s : set the default thread stack size (KB)\n", OPTION_THREAD_STACK_SIZE);
+  printf("%s : use sleep between node creation and memory lock\n", OPTION_USE_SLEEP);
+}
+
+bool configure_malloc_behavior()
+{
+  // Lock all current and future pages
+  if (mlockall(MCL_CURRENT | MCL_FUTURE) != 0) {
+    std::cerr << "mlockall failed. Error code " << strerror(errno) << std::endl;
+    return false;
+  }
+
+  // Turn off malloc trimming.
+  if (mallopt(M_TRIM_THRESHOLD, -1) == 0) {
+    std::cerr << "mallopt for trim threshold failed. Error code " << strerror(errno) << std::endl;
+    return false;
+  }
+
+  // Turn off mmap usage.
+  if (mallopt(M_MMAP_MAX, 0) == 0) {
+    std::cerr << "mallopt for mmap failed. Error code " << strerror(errno) << std::endl;
+    mallopt(M_TRIM_THRESHOLD, 128 * 1024);
+    munlockall();
+    return false;
+  }
+
+  return true;
+}
+
+bool reserve_process_memory(size_t memory_size)
+{
+  void * buf = nullptr;
+  const size_t pg_sz = sysconf(_SC_PAGESIZE);
+  int res;
+  res = posix_memalign(&buf, static_cast<size_t>(pg_sz), memory_size);
+  if (res != 0) {
+    std::cerr << "Failed to reserve memory. Error code " << strerror(errno) << std::endl;
+    return false;
+  }
+  memset(buf, 0, memory_size);
```

Comment:
So if `configure_memory_behavior` is called before this, this line is technically not needed as `mlockall(MCL_FUTURE)` will cause all allocated pages to be immediately mapped and locked.

I understand that this program allows the `configure_memory_behavior` to be turned off, which would necessitate this. There are two possible things we can do:

- Maybe always call `configure_memory_behavior` and don't have an option to turn it off. I say this because if `mlockall` is not called, the kernel can always swap your memory due to pressure, which would result in page fault during access later.
- At least put a comment saying this line is not needed if mlockall is called, and reference to the mlockall man page?

Reply:
Ok, I see your point now.


>Maybe always call configure_memory_behavior and don't have an option to turn it off. I say this because if mlockall is not called, the kernel can always swap your memory due to pressure, which would result in page fault during access later.

I agree, memory reserve should be called ONLY if `configure_memory_behavior` is called, otherwise is useless. 

>At least put a comment saying this line is not needed if mlockall is called, and reference to the mlockall man page?

I would remove it if it's not really needed.

Reply:
Sounds good. As a note, because I'm not a contributor in this repo, I can't resolve this conversation.

Reply:
I added you as a maintainer, see the invitation. Don't resolve this conversation yet, I have to remove the memset line first.

Reply:
Now we call `configure_memory_behavior` only if the memory has been locked